src/bitmap.rs

```diff
@@ -0,0 +1,85 @@
+//! This module holds abstractions that enable tracking the areas dirtied by writes of a specified
+//! length to a given offset. In particular, this is used to track write accesses within a
+//! `GuestMemoryRegion` object, and the resulting bitmaps can then be aggregated to build the
+//! global view for the entire `GuestMemory`.
+
+use std::fmt::{self, Debug};
+
+/// Common bitmap operations.
+// This is just an initial proposal; we'll add other methods if they need to be part of the main
+// interface as well.
+pub trait Bitmap: Default {
+    // For now this is mostly useful to preserve compatibility with the old `GuestMemoryMmap` (and
+    // potentially other backends) constructors, which are not bitmap-aware. We can remove this
+    // from the interface (and apply changes elsewhere) if it's too restrictive and/or configuring
+    // the bitmap post-creation (via `GuestMemoryRegion::bitmap_mut`) gets unwieldy.
+    /// Create a new bitmap of the given size.
+    fn new(size: usize) -> Self;
+
+    /// Mark the memory range specified by the given `offset` and `len` as dirtied.
+    fn mark_dirty(&self, offset: usize, len: usize);
```

Comment:
What would a `impl Bitmap for Foo` look like - more precisely, what would `Foo` contain to represent the bitmap? Mult be something with interior mutability so you can get away with `&self` here - are you thinking direct pointer manipulation? Splitting the bitmap representation into 64-bit chunks and using `Atomic`s? Something else?